client/controller/index.web.js

```diff
@@ -133,6 +133,35 @@ export function redirectLoggedOut( context, next ) {
 	next();
 }
 
+export function redirectLoggedOutForGifting( context, next ) {
+	const state = context.store.getState();
+	const userLoggedOut = ! isUserLoggedIn( state );
+
+	if ( userLoggedOut ) {
+		const loginParameters = {
+			redirectTo: context.path,
+		};
+
+		// Pass along "login_email" and "login_locale" parameters from the
+		// original URL, to ensure the login form is pre-filled with the
+		// correct email address and built with the correct language (when
+		// either of those are requested).
+		const login_email = getImmediateLoginEmail( state );
+		if ( login_email ) {
+			loginParameters.emailAddress = login_email;
```

Comment:
Noting that while mutation here is well-contained and pretty harmless, we should generally be avoiding it because it could end up being in places where it could cause subtle bugs.

Reply:
The thing here is I copied the `redirectLoggedOut` function 😮

Reply:
Oh, I see. In that case it might be a good opportunity to refactor in order to reuse and reduce duplication, WDYT?

Reply:
Sounds good, I will do it, and will use your suggestion below